File: src/main/java/fr/centralesupelec/sio/endpoints/AuthFilter.java
package fr.centralesupelec.sio.endpoints;

import javax.servlet.*;
import javax.servlet.annotation.WebFilter;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@WebFilter(urlPatterns = "/movies")
public class AuthFilter implements Filter {

    private Pattern HEADER_PATTERN = Pattern.compile("Bearer (.+)");

    @Override
    public void init(FilterConfig filterConfig) throws ServletException {

    }

    @Override
    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain) throws IOException, ServletException {
        String auth = ((HttpServletRequest) servletRequest).getHeader("Authorization");
        if (auth == null || auth.isEmpty()) {
            ApiServlet.writeError("Missing Authorization header", HttpServletResponse.SC_UNAUTHORIZED, (HttpServletResponse) servletResponse);
            return;
        }
        Matcher m = HEADER_PATTERN.matcher(auth);
        if (!m.matches()) {
            ApiServlet.writeError("Invalid Authorization header", HttpServletResponse.SC_UNAUTHORIZED, (HttpServletResponse) servletResponse);
            return;
        }
        if (!AuthManager.checkAccessToken(m.group(1))) {
            ApiServlet.writeError("Invalid access token", HttpServletResponse.SC_UNAUTHORIZED, (HttpServletResponse) servletResponse);
            return;
        }
        filterChain.doFilter(servletRequest, servletResponse);
    }

    @Override
    public void destroy() {

    }

}

File: src/main/java/fr/centralesupelec/sio/endpoints/AuthFilter.kt
package fr.centralesupelec.sio.endpoints

import fr.centralesupelec.sio.endpoints.utils.writeError
import java.io.IOException
import java.util.regex.Pattern
import javax.servlet.*
import javax.servlet.annotation.WebFilter
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

/**
 * A filter to check that all requests to the API are authenticated with a valid token.
 */
// TODO: Configure the filter to match ALL endpoints (except for the login endpoint)
@WebFilter(urlPatterns = ["/movies"])
class AuthFilter : Filter {

    private val headerPattern = Pattern.compile("Bearer (.+)")

    @Throws(ServletException::class)
    override fun init(filterConfig: FilterConfig?) {
    }

    @Throws(IOException::class, ServletException::class)
    override fun doFilter(servletRequest: ServletRequest, servletResponse: ServletResponse, filterChain: FilterChain) {

        val auth = (servletRequest as HttpServletRequest).getHeader("Authorization")
        if (auth.isNullOrEmpty()) {
            (servletResponse as HttpServletResponse).writeError(
                    "Missing Authorization header",
                    HttpServletResponse.SC_UNAUTHORIZED
            )
            return
        }

        val m = headerPattern.matcher(auth)
        if (!m.matches()) {
            (servletResponse as HttpServletResponse).writeError(
                    "Invalid Authorization header",
                    HttpServletResponse.SC_UNAUTHORIZED
            )
            return
        }

        if (!AuthManager.checkAccessToken(m.group(1))) {
            (servletResponse as HttpServletResponse).writeError(
                    "Invalid access token",
                    HttpServletResponse.SC_UNAUTHORIZED
            )
            return
        }

        filterChain.doFilter(servletRequest, servletResponse)
    }

    override fun destroy() {
    }

}
